syscall.go

```diff
@@ -178,17 +179,76 @@ func waitForBlockDevice(deviceName string) error {
 	return nil
 }
 
-func mountContainerRootFs(containerID, image, rootFs, fsType string) (string, error) {
+func recanSCSIBus() error {
+	scsiHostPath := "/sys/class/scsi_host/"
+	if _, err := os.Stat(scsiHostPath); err != nil {
+		return err
+	}
+
+	files, err := ioutil.ReadDir(scsiHostPath)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		host := file.Name()
+		scanPath := filepath.Join(scsiHostPath, host, "/scan")
+		data := []byte("- - -")
+		if err := ioutil.WriteFile(scanPath, data, 0666); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// findSCSIDisk finds the SCSI disk name associated with the given SCSI address
+// This approach eliminates the need to predict the disk name on the host side,
+// but we do need to rescan SCSI bus for this.
+func findSCSIDisk(scsiAddr string) (string, error) {
+	scsiPath := fmt.Sprintf("/sys/class/scsi_disk/0:0:%s/device/block/", scsiAddr)
+
+	if _, err := os.Stat(scsiPath); err != nil {
+		return "", err
+	}
+
+	files, err := ioutil.ReadDir(scsiPath)
+	if err != nil {
+		return "", err
+	}
+
+	if len(files) > 1 {
+		return "", fmt.Errorf("Expecting a single SCSI device, found %v", files)
+	}
+
+	return files[0].Name(), nil
+}
+
+func mountContainerRootFs(containerID, image, rootFs, fsType, scsiAddr string) (string, error) {
 	dest := filepath.Join(containerMountDest, containerID, "root")
 	if err := os.MkdirAll(dest, os.FileMode(0755)); err != nil {
 		return "", err
 	}
 
 	var source string
 	if fsType != "" {
-		source = filepath.Join(devPath, image)
-		if err := waitForBlockDevice(image); err != nil {
-			return "", err
+		// If SCSI adddress is provided, use that to find SCSI disk
+		if scsiAddr != "" {
```

Comment:
You should rely on `fsType` to get something like `"scsi"`. In case you determine this is a SCSI device, then you check the `expectedName`, rescan the bus and find the device, and after that you can reassign `fsType` to `bind`.

Reply:
I do not understand, fstype is supposed to container the file system type of the block device.
In case fstype is non-empty and scsi-addr is non-empty, we try to find the device name based on the scsi address sent. In case scsi-addr is empty but fstype is populated, we fall back to virtio-block and rely on the expected name passed.